keps/sig-storage/20191125-bucket-provisioning.md

```diff
@@ -0,0 +1,428 @@
+
+#### BucketAccessRequest
+
+A user facing API representing an object store user and an access policy defining the user’s relation to a storage instance.  A user will create a BucketAccessRequest in the app namespace.  A BucketAccessRequest can specify *either* a ServiceAccount in the same namespace or a desired Secret name.  Specifying a ServiceAccount enables provisioners to support cloud provider identity integration with their respective Kubernetes cluster offerings.
+
+```yaml
+apiVersion: cosi.io/v1alpha1
+kind: BucketAccessRequest
+metadata:
+  name:
+  namespace:
+  labels:
+    cosi.io/provisioner: [1]
+  finalizers:
+  - cosi.io/finalizer [2]
+spec:
+  serviceAccountName: [3]
+  bucketRequest: [4] 
+  bucketAccessClassName: [5]
+  bucketAccessName: [6]
+status:
+  message: [7]
+  phase: [8]
+```
+
+1. `labels`: should be added by controller.  Key’s value should be the provisioner name. Characters that do not adhere to [Kubernetes label conventions](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set) will be converted to ‘-’.
+1. `finalizers`: should be added by controller to defer `BucketAccessRequest` deletion until backend deletion ops succeed.
+1. `serviceAccountName`: the name of a Kubernetes ServiceAccount in the same namespace.  This field is included to support cloud provider identity integration.  Should not be set when specifying `accessSecretName`.
+1. `bucket`: The the name of the bucket to which the user identity or ServiceAccount should be granted access to, according to the policies defined in the `BucketAccessClass`.
+1. `bucketAccessClassName`: name of the `BucketAccessClass` specifying the desired set of policy actions to be set for a user identity or ServiceAccount.
+1. `bucketAccessName`: name of the bound cluster-scoped `BucketAccess` instance
+1. `message`: a human readable description detailing the reason for the current `phase``.
+1. `phase`: is the current state of the Bucket:
+   - _Bound_: the controller finished processing and has bound the BucketAccess to the BucketRequest via the BucketRequest.Spec.BucketAccessRequestName and BucketRequest.Spec.BucketAccessRequestNamespace
+   - _Failed_: error and all retries have been exhausted.
```

Comment:
When are _all retries exhausted_? I thought, in general, k8s will continue to retry and this is reflected in controller-runtime.

Reply:
Yeah, the states should be 

Created
Bound

and various
ErrXYZ states